### PR TITLE
Bitcoin Message Signing and Verification

### LINKED ISSUE #2597: Bitcoin message signing and verification
**Is your feature request related to a problem? Please describe.**
Not that I know of.  But I believe somebody thought about that at least once.

**Describe the solution you'd like**
It would be nice to have an ability to sign and verify messages without relying on third-party applications, such as bitcoin node or SPV clients. 

**Describe alternatives you've considered**
Currently I'm using OpenSSL-based solution to implement this functionality on application level. This is too far from optimal one and I'd be glad to get rid of OpenSSL dependency.

**Additional context**

There is a python implementation for reference:

https://github.com/stequald/bitcoin-sign-message/blob/master/signmessage.py

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Some Bitcoin wallets support a message signing & verification format, to create a proof (a signature) that someone has access to the private keys of a specific address.
This feature currently works on old legacy addresses only, and is not widely used, but we can add it nonetheless.

Note that C++ methods are added, not currently exposed over C interface.

Fixes #2597 .

## How to test

1. See unit tests in `MessageSignerTests`.
2. Generate some signature with the unit tests, and verify them with an online 3rd party verifier:
https://www.bitcoin.com/tools/verify-message/
https://www.verifybitcoinmessage.com/
https://reinproject.org/bitcoin-signature-tool/
3. Pick a private key, generate WIF format, generate corresponding legacy address.  Use one of the services above to create a signature. Validate with unit tests.
3. Using `bitcoin-cli`, generate a test wallet (non-descriptor based), get new address (legacy!), dump the private key, generate a signature. Validate that signature with unit tests.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

**Comment by hewigovens** [comment]
It's not useful at all without C interface IMHO

File: src/Bitcoin/MessageSigner.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+Data msgToHash(const std::string& msg) {
+    Data d = msgToFullData(msg);
+    return Hash::sha256d(d.data(), d.size());
+}
+
+std::string MessageSigner::signMessage(const PrivateKey& privateKey, const std::string& address, const std::string& message, bool compressed) {
+    if (!Address::isValid(address)) {
+        throw std::invalid_argument("Address is not valid (legacy) address");
+    }
+    std::string addrFromKey = "";
```
Comment (Milerius) [inline src/Bitcoin/MessageSigner.cpp:42]: std::string will be initialized empty

File: src/Bitcoin/MessageSigner.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    const auto messageHash = msgToHash(message);
+    auto recId = signature[0];
+    auto compressed = false;
+    if (recId >= 31) {
+        recId -= 4;
+        compressed = true;
+    }
+    if (recId >= 27) {
+        recId -= 27;
+    }
```
Comment (Milerius) [inline src/Bitcoin/MessageSigner.cpp:78]: Would be nice to name the constant value here.

File: src/PublicKey.cpp [hunk 3]
```diff
@@ -206,7 +206,25 @@ Data PublicKey::hash(const Data& prefix, Hash::Hasher hasher, bool skipTypeByte)
...
+    if (signatureRS.size() < 64) {
+        throw std::invalid_argument("signature too short");
+    }
+    if (recId >= 4) {
+        throw std::invalid_argument("Invalid recId (>=4)");
+    }
+    if (messageDigest.size() < 32) {
+        throw std::invalid_argument("digest too short");
+    }
+    TW::Data result(65);
```
Comment (Milerius) [inline src/PublicKey.cpp:219]: same remark regarding constant values

File: src/PublicKey.cpp [hunk 4]
```diff
@@ -206,7 +206,25 @@ Data PublicKey::hash(const Data& prefix, Hash::Hasher hasher, bool skipTypeByte)
...
+    int ret = ecdsa_recover_pub_from_sig(&secp256k1, result.data(), signatureRS.data(), messageDigest.data(), recId);
+    if (ret != 0) {
+        throw std::invalid_argument("recover failed " + std::to_string(ret));
+    }
+    return PublicKey(result, TWPublicKeyTypeSECP256k1Extended);
```
Comment (Milerius) [inline src/PublicKey.cpp:224]: Here you can use if initializer c++17 feature:

```cpp
if (int ret = statement; ret)
```

**Review by Milerius — COMMENTED** [review]
Good job, same remark as @hewigovens would be nice to have C interface and some example in swift/kotlin in the same pull request.

**Review by Milerius — DISMISSED** [review]

**Comment by catenocrypt** [comment]
Add C iface, with swift and C/C++ tests

**Review by Milerius — APPROVED** [review]

**Review by hewigovens — APPROVED** [review]

**Pull request merged**